File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        // Convert the string to lowercase and remove non-alphanumeric characters
        // This regular expression replaces any character that is not an alphabet (a-z, A-Z) or a number (0-9) with an empty string, effectively removing them from the original string
        s = s.replaceAll("[^a-zA-Z0-9]", "");
        s = s.toLowerCase();

        int left = 0;
        int right = s.length()-1;

        while(left < right){
            if(s.charAt(left) != s.charAt(right))
            {
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
}

/*
Character.toLowerCase(A)
Character.isLetterOrDigit(j)

*/

File: 0125-valid-palindrome/0125-valid-palindrome.go
func isPalindrome(s string) bool {
	runes := []rune(s) // convert string to slice of runes for safe indexing
	left, right := 0, len(runes)-1

	for left < right {
		l, r := runes[left], runes[right]

		// Skip non-alphanumeric from left
		if !unicode.IsLetter(l) && !unicode.IsDigit(l) {
			left++
			continue
		}

		// Skip non-alphanumeric from right
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) {
			right--
			continue
		}

		// Compare lowercase forms
		if unicode.ToLower(l) != unicode.ToLower(r) {
			return false
		}

		left++
		right--
	}

	return true
}